rllib/offline/json_reader.py

```diff
@@ -173,31 +176,45 @@ def _try_parse(self, line: str) -> Optional[SampleBatchType]:
         if not line:
             return None
         try:
-            batch = _from_json(line)
+            batch = self._from_json(line)
         except Exception:
             logger.exception("Ignoring corrupt json record in {}: {}".format(
                 self.cur_file, line))
             return None
 
         # Clip actions (from any values into env's bounds), if necessary.
         cfg = self.ioctx.config
-        if cfg.get("clip_actions"):
+        if cfg.get("clip_actions") and self.ioctx.worker is not None:
             if isinstance(batch, SampleBatch):
                 batch[SampleBatch.ACTIONS] = clip_action(
-                    batch[SampleBatch.ACTIONS], self.ioctx.worker.policy_map[
-                        "default_policy"].action_space_struct)
+                    batch[SampleBatch.ACTIONS],
+                    self.default_policy.action_space_struct)
             else:
                 for pid, b in batch.policy_batches.items():
                     b[SampleBatch.ACTIONS] = clip_action(
                         b[SampleBatch.ACTIONS],
                         self.ioctx.worker.policy_map[pid].action_space_struct)
-        # Re-normalize actions (from env's bounds to 0.0 centered), if
+        # Re-normalize actions (from env's bounds to zero-centered), if
         # necessary.
-        if cfg.get("actions_in_input_normalized") is False:
+        if cfg.get("actions_in_input_normalized") is False and \
+                self.ioctx.worker is not None:
             if isinstance(batch, SampleBatch):
-                batch[SampleBatch.ACTIONS] = normalize_action(
-                    batch[SampleBatch.ACTIONS], self.ioctx.worker.policy_map[
-                        "default_policy"].action_space_struct)
+                pol = self.default_policy
+                # If we have a complex action space and actions were flattened
+                # and we have to normalize -> Error.
+                if isinstance(pol.action_space_struct, (tuple, dict)) and \
+                        not pol.config.get("_disable_action_flattening"):
+                    raise ValueError(
+                        "Normalization of offline actions that are flattened "
+                        "is not supported! Make sure that you record actions "
+                        "into offline file with the "
+                        "`_disable_action_flattening=True` flag OR as already "
+                        "normalized (between -1.0 and 1.0) values. When "
+                        "reading already normalized action values, make sure "
+                        "to set `actions_in_input_normalized=True`.")
+                else:
+                    batch[SampleBatch.ACTIONS] = normalize_action(
+                        batch[SampleBatch.ACTIONS], pol.action_space_struct)
             else:
                 for pid, b in batch.policy_batches.items():
```

Comment:
a quick question, how come when we going through policy_batches, we don't have to worry about the error case that the offline actions are already flattened like above?
should we use the same logics to process these 2 cases?

Reply:
Great catch! Let me fix this. We should share the error message.

Reply:
I can't seem to find where the diff is for this fix.
but I am gonna assume we handle this now.
thanks.

Reply:
Oh, sorry, you are right. Will do this now before merging. Thanks for surfacing this again!